x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java

```diff
@@ -482,6 +501,41 @@ public void accessGranted(String requestId, Authentication authentication, Strin
                 logger.info(AUDIT_MARKER, logEntry);
             }
         }
+        // "config change" records are not filtered out by ignore policies (i.e. they are always printed).
+        // The security changes here are the consequences of *user* requests,
+        // so in a strict interpretation we should filter them out if there are ignore policies in place for the causing user,
+        // but we don't because filtering out security changes by the causing user is trappy.
+        if (events.contains(SECURITY_CONFIG_CHANGED)) {
+            String eventAction = null;
+            if (msg instanceof PutUserRequest ||
+                    msg instanceof PutRoleMappingRequest ||
+                    msg instanceof PutRoleRequest ||
+                    msg instanceof SetEnabledRequest ||
+                    msg instanceof ChangePasswordRequest ||
+                    msg instanceof CreateApiKeyRequest ||
+                    msg instanceof GrantApiKeyRequest ||
+                    msg instanceof CreateTokenRequest ||
+                    msg instanceof PutPrivilegesRequest) {
+                eventAction = "put";
+            } else if (msg instanceof DeleteUserRequest ||
+                    msg instanceof DeleteRoleMappingRequest ||
+                    msg instanceof DeleteRoleRequest ||
+                    msg instanceof InvalidateTokenRequest ||
+                    msg instanceof InvalidateApiKeyRequest ||
+                    msg instanceof DeletePrivilegesRequest) {
+                eventAction = "delete";
+            }
```

Comment:
Even though we've discussed this before, I think you're raising a good point.
I believe the root question is to what extent is the auditing different from generic request serialisation. Given that the present audit approach is not designed around the concept of "requests" but rather centred on "changes", I think I can see the case for an audit-specific serialisation logic.

Here's a list for why I think that is the case:
* the refresh policy is not important for auditing
* the name of the changed object (role, user, etc) should be distinguished as a separate field, not as part of a serialised object
* the nature of the change (put, enable, clear) should also by distinguished as a separate field

Reply:
@ywangd After mulling over it some more, I've changed back to thinking that the code to print the request bodies should sit in the `LoggingAuditTrail` class.

It's probably easier to show it in code than to talk about, but the reasoning is that the layout of it is particular to the logging audit trail format, and it's not reasonable to expect the reuse of it. It's also easier to ensure and maintain the formatting consistency, if it all sits together rather than be scattered across request methods. A simple example, is the consistency for the `username` vs `user.name` fields.

Reply:
I guess I'll have to see the code to fully understand what you mean by "the code to print the request bodies". But I didn't suggest to move the "print" part to each individual class. The actual print should always be done in `LoggingAuditTrail` and I agree with it. I was referring to the "serialization" part of the logging message, i.e. the `toXContent` method, where the message is prepared. Also on the topic of resuing the auditing message, there were talks about auditing into an index (or maybe some other connectors). Would it be a valid possible future reuse?

Reply:
My point is that even the `toXContent` is not really suitable as a request object method.
The JSON object for the config change is part of the JSON object for the whole event so they must use the same  conventions, eg for field names (`username` vs `user.name`), or plurals, or tenses (past/present). The objects must also be consistent among themselves. For example, check out the handling of the `SetEnabledRequest` and `GrantApiKeyRequest` requests.
In general, I've chosen the field names in the context of the audit log that we've got; most commonly field names coincide with the XContent field names, but not always. Same for the nesting patterns.

> Also on the topic of resuing the auditing message, there were talks about auditing into an index (or maybe some other connectors). Would it be a valid possible future reuse?

If, as initially planned, this feature would've been a particular case of "request body auditing" , then yes that would be a valid reuse case. But it's now a different thing. Audit records are a richer translation of requests in a friendlier format in the particular context of security config changes to be applied.

Slightly off-topic, the index connector is not a suitable connector for audits as the buffering on the local disk is really required (we had something like that and we've deprecated it).